Sources/Sentry/SentryANRTrackingIntegration.m

```diff
@@ -195,15 +195,85 @@ - (void)anrStoppedWithResult:(SentryANRStoppedResult *_Nullable)result
     [self.fileManager deleteAppHangEvent];
 
     // We round to 0.1 seconds accuracy because we can't precicely measure the app hand duration.
-    NSString *errorMessage =
-        [NSString stringWithFormat:@"App hanging between %.1f and %.1f seconds.",
-            result.minDuration, result.maxDuration];
-
+    NSString *appHangDurationInfo = [NSString
+        stringWithFormat:@"between %.1f and %.1f seconds", result.minDuration, result.maxDuration];
+    NSString *errorMessage = [NSString stringWithFormat:@"App hanging %@.", appHangDurationInfo];
     event.exceptions.firstObject.value = errorMessage;
-    [SentrySDK captureEvent:event];
+
+    NSMutableDictionary *mechanismData = [event.exceptions.firstObject.mechanism.data mutableCopy];
+    [mechanismData removeObjectForKey:SentryANRMechanismDataAppHangDuration];
+    event.exceptions.firstObject.mechanism.data = mechanismData;
+
+    // We already applied the scope. We use an empty scope to avoid overwriting exising fields on
+    // the event.
+    [SentrySDK captureEvent:event withScope:[[SentryScope alloc] init]];
 #endif // SENTRY_HAS_UIKIT
 }
 
+- (void)captureStoredAppHangEvent
+{
+    __weak SentryANRTrackingIntegration *weakSelf = self;
+    [self.dispatchQueueWrapper dispatchAsyncWithBlock:^{
+        if (weakSelf == nil) {
+            return;
+        }
+
+        SentryEvent *event = [weakSelf.fileManager readAppHangEvent];
+        if (event == nil) {
+            return;
+        }
+
+        [weakSelf.fileManager deleteAppHangEvent];
+
+        if (weakSelf.crashWrapper.crashedLastLaunch) {
+            // The app crashed during an ongoing app hang. Capture the stored app hang as it is.
+            // We already applied the scope. We use an empty scope to avoid overwriting exising
+            // fields on the event.
+            [SentrySDK captureEvent:event withScope:[[SentryScope alloc] init]];
+        } else {
+            // Fatal App Hang
+            // We can't differ if the watchdog or the user terminated the app, because when the main
+            // thread is blocked we don't receive the applicationWillTerminate notification. Further
+            // investigations are required to validate if we somehow can differ between watchdog or
+            // user terminations; see https://github.com/getsentry/sentry-cocoa/issues/4845.
+
+            if (event.exceptions.count != 1) {
```

Comment:
`l`: how would it behave with multiple exceptions? can that even happen?

Reply:
That's unexpected. If that's the case, something is clearly wrong, as explained with the warn log message we rather drop the event.